sklearn/cross_decomposition/_pls.py

```diff
@@ -261,101 +265,151 @@ def fit(self, X, Y):
-            y_scores = np.dot(Yk, y_weights) / y_ss
-
-            # Deflation: subtract rank-one approx to obtain Xk+1 and Yk+1
-            x_loadings = np.dot(x_scores, Xk) / np.dot(x_scores, x_scores)
-            Xk -= np.outer(x_scores, x_loadings)
-
-            if self.deflation_mode == "canonical":
-                # regress Yk on y_score
-                y_loadings = np.dot(y_scores, Yk) / np.dot(y_scores, y_scores)
-                Yk -= np.outer(y_scores, y_loadings)
-            if self.deflation_mode == "regression":
-                # regress Yk on x_score
-                y_loadings = np.dot(x_scores, Yk) / np.dot(x_scores, x_scores)
-                Yk -= np.outer(x_scores, y_loadings)
-
-            self.x_weights_[:, k] = x_weights
-            self.y_weights_[:, k] = y_weights
-            self._x_scores[:, k] = x_scores
-            self._y_scores[:, k] = y_scores
-            self.x_loadings_[:, k] = x_loadings
-            self.y_loadings_[:, k] = y_loadings
-
-        # X was approximated as Xi . Gamma.T + X_(R+1)
-        # Xi . Gamma.T is a sum of n_components rank-1 matrices. X_(R+1) is
-        # whatever is left to fully reconstruct X, and can be 0 if X is of rank
-        # n_components.
-        # Similarly, Y was approximated as Omega . Delta.T + Y_(R+1)
-
-        # Compute transformation matrices (rotations_). See User Guide.
-        self.x_rotations_ = np.dot(
-            self.x_weights_,
-            pinv2(np.dot(self.x_loadings_.T, self.x_weights_), check_finite=False),
-        )
-        self.y_rotations_ = np.dot(
-            self.y_weights_,
-            pinv2(np.dot(self.y_loadings_.T, self.y_weights_), check_finite=False),
-        )
-        # TODO(1.3): change `self._coef_` to `self.coef_`
-        self._coef_ = np.dot(self.x_rotations_, self.y_loadings_.T)
-        self._coef_ = (self._coef_ * self._y_std).T
-        self.intercept_ = self._y_mean
-        self._n_features_out = self.x_rotations_.shape[1]
+                t = Xk @ r
+                tt = t.T @ t
+                p = (Xk.T @ t) / tt
+                q = (r.T @ S) / tt
+                # in the case of univariate Y we can simply do p * q, but what
+                # we want in general is np.outer(p, q). p[:, None] * q is faster
+                # than np.outer(p, q) and still generalizes
+                S = S - tt * p[:, None] * q
+
+                self.x_weights_[:, k] = w.ravel()
+                self.x_rotations_[:, k] = r.ravel()
+                self.x_loadings_[:, k] = p.ravel()
+                self.y_loadings_[:, k] = q
+
+            self._coef_ = ((self.x_rotations_ @ self.y_loadings_.T) * self._y_std).T
+            self.intercept_ = self._y_mean
+            self._n_features_out = self.x_rotations_.shape[1]
+            self._x_scores = Xk @ self.x_rotations_
+            self._y_scores = None
+            self.y_weights_ = None
+            self.y_rotations_ = None
+
```

Comment:
It would be nice to factorize those 2 blocks of code in appropriately named private methods.

Reply:
I broke out the NIPALS and Dayal-MacGregor methods into two separate private methods